vault/resource_approle_auth_backend_login.go

```diff
@@ -0,0 +1,209 @@
+			},
+			"lease_duration": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "How long the token is valid for.",
+			},
+			"lease_started": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The timestamp the lease started on, as determined by the machine running Terraform.",
+			},
+			"accessor": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The accessor for the token.",
+			},
+			"client_token": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The token.",
+			},
+			"metadata": {
+				Type:        schema.TypeMap,
+				Computed:    true,
+				Description: "Metadata associated with the token.",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"backend": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Unique name of the auth backend to configure.",
+				ForceNew:    true,
+				Default:     "approle",
+				// standardise on no beginning or trailing slashes
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+		},
+	}
+}
+
+func approleAuthBackendLoginCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	backend := d.Get("backend").(string)
+
+	path := approleAuthBackendLoginPath(backend)
+
+	log.Printf("[DEBUG] Logging in with AppRole auth backend %q", path)
+	data := map[string]interface{}{
+		"role_id": d.Get("role_id").(string),
+	}
+	if v, ok := d.GetOk("secret_id"); ok {
+		data["secret_id"] = v.(string)
+	}
+
+	resp, err := client.Logical().Write(path, data)
+	if err != nil {
+		return fmt.Errorf("Error logging into AppRole auth backend %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Logged in with AppRole auth backend %q", path)
+
+	d.SetId(resp.Auth.Accessor)
+	d.Set("lease_started", time.Now().Format(time.RFC3339))
+	d.Set("client_token", resp.Auth.ClientToken)
+
+	return approleAuthBackendLoginRead(d, meta)
+}
+
+func approleAuthBackendLoginRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	log.Printf("[DEBUG] Reading token %q", d.Id())
+	resp, err := client.Auth().Token().LookupAccessor(d.Id())
+	if err != nil {
+		return fmt.Errorf("Error reading token %q from Vault: %s", d.Id(), err)
+	}
+	if resp == nil {
+		log.Printf("[DEBUG] Token %q not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+	log.Printf("[DEBUG] Read token %q", d.Id())
+	if leaseExpiringSoon(d, client) {
+		log.Printf("[DEBUG] Lease for %q expiring soon, renewing", d.Id())
+		renewed, err := client.Auth().Token().Renew(d.Get("client_token").(string), d.Get("lease_duration").(int))
+		if err != nil {
+			log.Printf("[DEBUG] Error renewing token %q, bailing", d.Id())
+		} else {
+			resp = renewed
+			d.Set("lease_started", time.Now().Format(time.RFC3339))
+			d.Set("client_token", resp.Auth.ClientToken)
+			d.SetId(resp.Auth.Accessor)
```

Comment:
Does this `resp.Auth.Accessor` value change? It's unusual to see `d.SetId` outside of a `create` method

Reply:
It should be a permanent, stable ID, but I think I was just setting it just to be safe. I'll remove it.

Reply:
I spoke too soon, sorry. Yes, this value changes, because we're essentially replacing the resource. I'm trading in one token (that has expired) for a new token (that has not). I dunno if it's cool to play around with resource IDs like that, but this resource is of minimal utility if I can't swap out the tokens. :/

Reply:
It's unusual to swap them, yes, but the circumstances seem to warrant it